### DISCUSSION TITLE
How To Activate spindle linearization and setting the piecewise ?

### DISCUSSION THREAD

**Discussion by hanifhan** [grblHAL/core#269] [Q&A] [answered]
my RPM doesnt seem accurate when i use an tachometer , so i want to try using spindle linearization , in previous GRBL i can use it with enabling piecewise , but how to activate it in GRBLHal ?

**Comment by terjeio** [comment, marked as answer]
It is enabled [here](https://github.com/grblHAL/core/blob/20aa4ec941bca158e2ded42351d9230085e2fbf2/grbl.h#L180).  Enabling givess access to settings for the values: $66 - $69. These are a bit odd as they each holds a comma separated list of 3 float values:
RPM, RPM_LINE_A. RPM_LINE_B
or a single `0` or a blank line to disable the entry.

I need to revisit this and document the settings...

**Reply by hanifhan** [reply]
thanks for your answer , i have activate it before , but i cant find the $66-$69 setting , where should i fill it ?

**Reply by terjeio** [reply]
Which driver?
It seems there are several parts that are broken for these settings...

**Reply by hanifhan** [reply]
im using stm32f446xx driver , 
but still , i dont know how to fill the piecewise setting in this GRBLHal , 
Thanks in advance for your response

**Reply by terjeio** [reply]
[Fix committed.](https://github.com/grblHAL/core/blob/master/changelog.md#20230317)

> but i cant find the $66-$69 setting , where should i fill it ?

In [ioSender](https://github.com/terjeio/ioSender) you can find the settings under the _Spindle_ grouping:

![image](https://user-images.githubusercontent.com/20260062/225825324-07b0f849-dd8c-4e7e-be49-4df478fb2744.png)

Other senders migth not show these settings, then you have to set the values from the MDI or console. Another fallback will be to connect with a terminal program such as puTTY. The final fallback will be to [change the default values in config.h](https://github.com/grblHAL/core/blob/cad1dc8476b9bdc8e751d2535dae7019f701a835/config.h#L1078-L1140) and recompile/reflash.

FYI I have moved the [enable setting to config.h](https://github.com/grblHAL/core/blob/cad1dc8476b9bdc8e751d2535dae7019f701a835/config.h#L96-L110).

**Reply by hanifhan** [reply]
hi thanks its help me a lot 
i've try that settings , but i think the computation still have a problem
i get the piecewise by https://github.com/terjeio/grblHAL/blob/master/doc/script/fit_nonlinear_spindle.py
and of course i have change the PWM Point 1 -3 to 30,50, and 70 
since the SPINDLE_PWM_MAX_VALUE is 100
but the result is my maximum pwm voltage only 0,9 volt , if i disable the linearisation my maximum pwm voltage normal in 9.9 v

**Reply by terjeio** [reply]
Use the RPM values. SPINDLE_PWM_MAX_VALUE  is a percentage and relates to the max RPM setting ($36).

**Reply by hanifhan** [reply]
i do my piecewise like i sent  the picture below , did i do wrong ?

**Comment by hanifhan** [comment]
![image](https://user-images.githubusercontent.com/78627497/226238587-d00b6420-3688-46b7-9540-eb9ef8f3f267.png)
i do my piecewise like that, where did i do wrong ? , because the output PWM only became 0,89 volt if i activate and set the linearization

**Reply by terjeio** [reply]
What is your RPM max setting ($30)? Should be 6887.7.

**Reply by hanifhan** [reply]
yea , its 6087,7

**Reply by terjeio** [reply]
Ok, post your $66-69 settings for me to test with.

**Reply by hanifhan** [reply]
![image](https://user-images.githubusercontent.com/78627497/226262975-d43567f5-5d65-435b-b89d-e45b2157e2d6.png)
here is my updated data speed
$30=6023.2 (rpm max)
$31=614.0 (rpm min)
$66=6.140000e+02,4.131690e-02,3.685755e-01 (piece 1)
$67=1.219079e+03,4.160557e-02,7.204920e-01 (piece 2)
$68=3.021722e+03,4.094967e-02,3.652996e-01 (piece 3)
$69=4.853239e+03,4.094967e-02,3.652996e-01 (piece 4)

**Reply by terjeio** [reply]
grblHAL drivers use at least a 16 bit period register for PWM, legacy Grbl a 8 bit register. This means that using a `PWM_set` array ending at 255 will not work. To complicate this further the different drivers does not utilize the full 16 bit range and may have different max values - based on the configured PWM frequency and the PWM timer source clock and possibly also a prescaler value that might change with the configured PWM frequency.

I'll have to recap a bit more to come up with a solution - best would be to calculate the correct PWM value based on a fixed range (0-1000?) and scale that in the firmware. An alternative would be to add a $ command to output the actual max value used internally. I wrote a program a few years back that did everything automatically, based on RPM readout from the controller, perhaps this has some clues...

**Reply by terjeio** [reply]
Try with setting $30 to 6700 and $31 to 0 and blank out $66-69. Then measure actual RPM vs programmed RPM (with the S value) and fill out the `PWM_set` array with programmed RPM values and `RPM_measured` array with the corresponding values.

You may initially set $30 to 1000, set the RPM to 1000 with M3S1000 and measure the RPM - then set $30 to the measured value before you run the linearization measurement sequence.

IMO the `PWM_set` variable name is misleading - it should be `RPM_set`...

**Reply by hanifhan** [reply]
i think  ever try that before , but it still doesnt work

**Reply by terjeio** [reply]
Ok - send me the settings from that try.

**Reply by hanifhan** [reply]
![image](https://user-images.githubusercontent.com/78627497/226302886-46ca5d80-d161-4cc4-bae4-287ca9f61831.png)
[Update these #define values and uncomment]
[ENABLE_PIECEWISE_LINEAR_SPINDLE in config.h.]
#define N_PIECES 4
#define RPM_MAX 1004.7
#define RPM_MIN 110.0
#define DEFAULT_RPM_POINT12 215.8
#define DEFAULT_RPM_POINT23 508.5
#define DEFAULT_RPM_POINT34 721.1
#define DEFAULT_RPM_LINE_A1 9.453451e-01
#define DEFAULT_RPM_LINE_B1 3.987956e+00
#define DEFAULT_RPM_LINE_A2 1.024824e+00
#define DEFAULT_RPM_LINE_B2 2.113793e+01
#define DEFAULT_RPM_LINE_A3 9.407338e-01
#define DEFAULT_RPM_LINE_B3 -2.162294e+01
#define DEFAULT_RPM_LINE_A4 1.057952e+00
#define DEFAULT_RPM_LINE_B4 6.290506e+01

[To operate over full model range, manually write these]
['$' settings or alter values in defaults.h. Grbl will]
[operate between min($30,RPM_MAX) and max($31,RPM_MIN)]
$30=1004.7 (rpm max)
$31=110.0 (rpm min)
$66=1.100000e+02,9.453451e-01,3.987956e+00 (piece 1)
$67=2.157815e+02,1.024824e+00,2.113793e+01 (piece 2)
$68=5.085148e+02,9.407338e-01,3.968503e+00 (piece 3)
$69=7.211148e+02,9.407338e-01,3.968503e+00 (piece 4)

**Reply by hanifhan** [reply]
i try using 1K RPM

**Reply by terjeio** [reply]
Are you compiling locally or are you using the Web Builder?

**Reply by hanifhan** [reply]
Using replit

**Reply by terjeio** [reply]
> Using replit

I do not know what "replit" is, assuming you are compiling locally then this line should be changed from:
https://github.com/grblHAL/core/blob/cad1dc8476b9bdc8e751d2535dae7019f701a835/spindle_control.c#L718
to:
`                   pwm_value = floorf((pwm_data->piece[idx].start * rpm - pwm_data->piece[idx].end) * pwm_data->pwm_gradient);`

and $30/$31 set to the actual min/max before collecting the values. $30/$31 should not be changed after collecting as this will change the gradient value. An example that seems to work correctly (with a very much linear profile):

```
CONFIG:
  N_pieces: 4
  PWM_min: 0.0
  PWM_max: 3000.0
  PWM_point1: 100.0
  PWM_point2: 1400.0
  PWM_point3: 2800.0
  N_data: 21
  PWM_set:  [   0.  150.  300.  450.  600.  750.  900. 1050. 1200. 1350. 1500. 1650.
 1800. 1950. 2100. 2250. 2400. 2550. 2700. 2850. 3000.]
  RPM_measured:  [   0.  147.  297.  447.  597.  747.  897. 1047. 1197. 1347. 1497. 1647.
 1797. 1947. 2097. 2247. 2397. 2547. 2697. 2847. 2997.]

SOLUTION:

[Update these #define values and uncomment]
[ENABLE_PIECEWISE_LINEAR_SPINDLE in config.h.]
#define N_PIECES 4
#define RPM_MAX 2997.0
#define RPM_MIN 0.0
#define RPM_POINT12 97.0
#define RPM_POINT23 1397.0
#define RPM_POINT34 2797.0
#define RPM_LINE_A1 1.030928e+00
#define RPM_LINE_B1 3.566591e-20
#define RPM_LINE_A2 1.000000e+00
#define RPM_LINE_B2 -3.000000e+00
#define RPM_LINE_A3 1.000000e+00
#define RPM_LINE_B3 -3.000000e+00
#define RPM_LINE_A4 1.000000e+00
#define RPM_LINE_B4 -3.000000e+00

[To operate over full model range, manually write these]
['$' settings or alter values in defaults.h. Grbl will]
[operate between min($30,RPM_MAX) and max($31,RPM_MIN)]
$30=2997.0 (rpm max)
$31=0.0 (rpm min)
$66=3.459593e-20,1.030928e+00,3.566591e-20 (piece 1)
$67=9.700000e+01,1.000000e+00,-3.000000e+00 (piece 2)
$68=1.397000e+03,1.000000e+00,3.459593e-20 (piece 3)
$69=2.797000e+03,1.000000e+00,3.459593e-20 (piece 4)

[No cpu_map.h changes required.]
```

**Reply by hanifhan** [reply]
wow , i've changed the code as ur command and its work perfectly ,thanks a lot :))
